### ISSUE TITLE
MBP + SceneGraph need support adding / removing geometry during run time. 

### ISSUE THREAD

**Issue by siyuanfeng-tri** [RobotLocomotion/drake#9665]
A lot of the planning applications rely on being able to add new or remove geometries on the fly. 
If the planner is also optimizing using dynamics, adding / removing rigid bodies on the fly also becomes highly desirable.

**Comment by amcastro-tri** [comment]
@edrumwri and @avalenzu are working on a composite MBP+SG, lets call it `MultibodyWorld` just for the sake of this discussion. My understanding is that there is a need of a workflow like this:
```
MultibodyWorld world;
AddModelFromSdfFile("my_model.sdf", &world.plant(), &world.scene_graph());
world.Finalize();
 // use the model of the world for something ....
auto context = world.CreateDefaultContext();

// Our perception system realized we have a new object in the world,
// add it so we can plan it.
const auto& body = world.AddRigidBody(...params...);
world.RegisterCollisionGeometry(body, Sphere(radius));
world.Finalize();  // must allow re-finalization. It could happen at CreateDefaultContext() time.

// Use the updated model...
auto context = world.CreateDefaultContext();

```

Please correct me if I'm wrong @siyuanfeng-tri.

**Comment by SeanCurtis-TRI** [comment]
@amcastro-tri and do you have an idea what it would look like to remove?

SG originally had an API for removing geometry. I can bring it back. It went away because the one thing we couldn't do is have it add/remove during the run of simulation in response to events. But the workflow that @amcastro-tri has sketched brings the logic for updating the world *outside* of the `Simulator` purview and places the burden on the programmer -- that gives us more flexibility.

So, @siyuanfeng-tri could you follow up on @amcastro-tri's example and indicate if that type of workflow works for you and, if not, propose an alternate?  Or elaborate if it's missing a nuance that you can anticipate.

**Comment by siyuanfeng-tri** [comment]
the example workflow is fairly accurate. the most wanted is being able to add geometry to exisiting bodies, adding bodies can come later. 
i also need to remove geometries on the fly, or being able to clone the dual.

**Comment by siyuanfeng-tri** [comment]
After thinking about it more, if it's not too much work, can you support adding / removing rigid bodies as well? It makes writing some constraints more straight forward. 
cc @hongkai-dai @avalenzu @calderpg-tri anything else?

**Comment by EricCousineau-TRI** [comment]
Just curious: Is there any use in making "un-finalizing" only possible via cloning? This way, any contexts / topology-depend stuff allocated beforehand are still considered valid?

Caveat: Storing any references to children elements (frames, bodies, etc.) would no longer be valid for the new instance.

**Comment by siyuanfeng-tri** [comment]
that's one way to do it. there is some concern about run time cost.

**Comment by SeanCurtis-TRI** [comment]
So are you doing modifications on the contents of the "world" with high frequency? And how big is the world? As we consider how we're spreading around the dynamics resources, this is a problem that has a *clunky* solution (i.e., a way to effectively accomplish the same thing) that can use current functionality -- this would obviously free up resources to tackle things that can't be done at all. However, if computation cost renders that a non-starter, it's would be important to know.

**Comment by tri-ltyyu** [comment]
Use cases (To Do: Sean to edit): 
1) Weld geometry to a frame 
2) Ability to set collision group of new geometry added

Acceptance criteria:
Need function to add geometry and query geometry for deletion. API to modify Context

**Comment by SeanCurtis-TRI** [comment]
The nature of this PR encompasses a larger set of features (including MBP). I propose that a new "sub-issue" be created capturing just the SceneGraph geometry/addition removal. We can defer the adding frames/bodies to another issue (also linked to this).

All in favor?

**Comment by sherm1** [comment]
Aye!

**Comment by siyuanfeng-tri** [comment]
aye!

**Comment by EricCousineau-TRI** [comment]
> After registering a new geometry, collision filtering can be applied to the new geometry and pre-existing geometries via SceneGraph::ExcludeCollisionsBetween()

To clarify on this and other bits, the `SceneGraph` people will be working with will be mutable? (So it'll be a tad bit odd in that you're fiddling with a mutable object that's actually registered within a diagram, most likely?)

**Comment by SeanCurtis-TRI** [comment]
@EricCousineau-TRI 

> the SceneGraph people will be working with will be mutable? 

I'm not sure what this means. The `Context` will be mutable. All of the geometry *lives* in the Context (in an abstract state variable). So, I'm unsure what your concern is. 

Why does playing with collision filters awake a concern that actually changing the population doesn't? Or do the "other bits" to which you allude capture both cases? In the API shown above, the `System` won't change. And the abstract state won't change in any way that affects the System connections. (It's a different story when we change the *frames*...but that's for a later day.)

**Comment by SeanCurtis-TRI** [comment]
@EricCousineau-TRI Good catch -- I've updated the API description to include a version of `ExcludeCollisionsBetween` that takes a Context. And, generally, all of those methods that define the world would have two versions: with and without a context.

**Comment by calderpg** [comment]
Sounds good to me. It sounds like `source_id` is something a MBP+SG wrapper would (and should) hide from the user.

**Comment by SeanCurtis-TRI** [comment]
@calderpg-tri and @siyuanfeng-tri This issue was tagged with *high* priority. From the conversation we had yesterday, I'd invite you to revisit the priority marking and updated it accordingly.

**Comment by siyuanfeng-tri** [comment]
The priority is somewhat arbitrary. How about I rephrase it to "available mid Nov"? I'd like to start porting anzu code to this + @hongkai-dai's new mbp + sg backend ik around that time.

**Comment by SeanCurtis-TRI** [comment]
I love the idea of indicating time frame. Could you document that in #9909 for the SG-specific functionality?

**Comment by siyuanfeng-tri** [comment]
I added a comment in #9909, not sure if hongkai's proj have a number that I can reference. Feel free to edit. also cc @avalenzu and @calderpg @tri-ltyyu

**Comment by EricCousineau-TRI** [comment]
Should we place this on a milestone that has dates? We had done it in Anzu for some bits, and that seemed to work decently well.

**Comment by hongkai-dai** [comment]
@siyuanfeng-tri , the PR I am working on is #9824

**Comment by tri-ltyyu** [comment]
This is supported for SG. Adding/removing geometry with mass for dynamics should be called out in new issue if needed, @siyuanfeng-tri. Closing this issue for now.
